### PR TITLE
Handle an indentation corner case for multi-line mirrors

### PR DESCRIPTION
**Pull request by mbriggs**
As requested, here is a contribution for the indentation issue raised in #665. It is a bit of a quick and dirty fix, but I don't think it is too bad.

### CODE REVIEW

**Comment by capitaomorte** [comment]
Nice start. I'm afraid it will break some test cases, such as when there are fields, particularly zero-length fields, immediately adjacent to the mirror, before or after. But anyway, I'm glad you contributed, and I can take it from here probably. And this being a <10 line contribution, I doubt you need an assignment. Thoughts on this @npostavs?

**Comment by mbriggs** [comment]
I ran the tests and nothing broke. Its a bit of a cludge to work it into the mirror update method, but ¯\_(ツ)_/¯

**Comment by capitaomorte** [comment]
FTR, fails for this snippet
```
(defun thingy ($1)
   (let ($2${1:$(mapconcat #'identity (split-string yas-text) "\n")})))
```
All goes nice and dandy at first, you input `bla ble bli` and get

```
(defun thingy (bla ble bli)
  (let (bla
        ble
        bli)))
```

Nice! Then you tab to `$2` and input `blo`. Ooops. This should be made into a test. 

@mbriggs I think the problem here is that `indent-region` doesn't preserve the positions of the markers of the zero-length fields. You can see what's happening if you try out this snippet (or an equivalent ruby one), with "C-u C-c C-t". 

We would need to write a `(yas--indent-region-preserving BEG END SNIPPET)`. I believe it can be extracted from existing logic around

```
(defun yas--indent-according-to-mode (snippet-markers)
  "Indent current line according to mode, preserving SNIPPET-MARKERS."...
``` 

in `yasnippet.el`. This is mostly a note to self, unless you @mbriggs, who have proven to like a challenge :-) decide to tackle it.

**Comment by capitaomorte** [comment]
@mbriggs I know it passed the current tests and Travis also runs them automatically for us. But as you know, passing automated tests is not a guarantee of quality.

It's not a particularly big kludge to make it in the mirror update method, the fact that `indent-region` is allowed to insert whitespace in misterious ways is what I have to worry about.

Nice blog, btw. Like your post about senior developers not really being senior.

File: yasnippet.el [hunk 1]
```diff
@@ -4240,7 +4240,13 @@ When multiple expressions are found, only the last one counts."
...
       (goto-char (yas--mirror-start mirror))
       (let ((yas--inhibit-overlay-hooks t))
-        (insert reflection))
+        (insert reflection)
+        (let ((start (yas--mirror-start mirror))
+              (end (yas--mirror-end mirror)))
+          (when (and (eq yas-indent-line 'auto)
+                     (not (eq (line-number-at-pos start)
+                              (line-number-at-pos end))))
+            (indent-region start end))))
```
Comment (npostavs) [inline yasnippet.el]: We either don't care about indenting the first line in which case we can just indent all of the lines unconditionally. Or else we do care, and we should make sure only to indent the last n-1 lines.

**Comment by npostavs** [comment]
> And this being a <10 line contribution, I doubt you need an assignment.

The official cutoff for "trivial" contributions is 15 lines (but it's cumulative).

**Comment by mbriggs** [comment]
FWIW I dont mind doing a copyright assignment, I just have some history with the FSF losing stuff, a previous boss not wanting to do it again, and ending up screwing over @magnars, who I respect quite a bit and still feel shitty about. Obviously that wouldnt happen here given that you are asking for it up front, but that experience left a bad taste in my mouth, and I dont really want to go down that road again for a small amount of work. 

The only reason I brought it up in the previous issue was cause it ALSO made me feel really shitty to not put in a bit of time to scratch an itch on an open source project I love and use a hundred times a day. Unfortunately I think I just came off as being a jerk, which wasn't the intention.

All that being said, I think at this point I *do* have a bit more investment in this issue now after all the back and forth, and I don't mind doing the copyright assignment and finishing this off. I am currently the CTO where I work, so getting that done shouldnt be a big deal.

**Comment by capitaomorte** [comment]
> Unfortunately I think I just came off as being a jerk

No, that's quite far from the truth, at least the way I read it.

Your contribution so far is trivial, and so it doesn't need a copyright assignment. It's the new function that needs to be written that would be non-trivial. Anyway, contribution is not just code, it's the report, the interaction, the fact that you made me think about it and influenced the design. So if you want to stop here, no prob. Otherwise get working on those forms, again I think you will find it's easier than you think.

**Comment by mbriggs** [comment]
quick update, email part of it has happened, waiting on them to get back to me about additional forms

**Comment by npostavs** [comment]
Indenting of multiline mirrors is now fixed in `master`, see 14f830a37dd146fb0671f20119d968837c4a8b11...eca2f995c15dda7632701f7cb682643f6c9597b6

**Pull request closed without merging**